gameSource/hetuwmod.cpp

```diff
@@ -1730,11 +1730,60 @@ void HetuwMod::SayStep() {
 void HetuwMod::Say(const char *text) {
 	if (bTeachLanguage) bTeachLanguage = false;
 
-	char *msg = new char[strlen(text)+1];
-	strcpy(msg, text);
+	char *msg = new char[strlen(text)*2+1];
+	encodeDigits(text, msg);
 	sayBuffer.push_back(msg);
 }
 
+// Encode digits using 0 = ?A, 1 = ?B, 2 = ?C, etc. due to server restrictions
+void HetuwMod::encodeDigits(const char *plain, char *encoded) {
+	bool questionMark = false;
+	int j = 0;
+	size_t len = strlen(plain);
+	for (size_t i=0; i<len; i++) {
+		if (isdigit(plain[i])) {
+			if (!questionMark) {
+				questionMark = true;
+				encoded[j++] = '?';
+			}
+			encoded[j++] = 'A' + plain[i] - '0';
+		} else {
+			questionMark = false;
+			encoded[j++] = plain[i];
+		}
+	}
+	encoded[j] = '\0';
+}
+
+void HetuwMod::decodeDigits(const char *encoded, char *plain) {
```

Comment:
It's more idiomatic for an in-place string mangling function like this to take one `char *` parameter. Otherwise, it's not obvious that these pointers alias in reality, let alone that they're always the same. Being able to also operate with a separate buffer is neat, but not worth any extra complexity/maintenance since a caller can just copy the buffer ahead of time at equal cost.